File: examples/python/example-actor-messaging.py
# SPDX-License-Identifier: BSL-1.0

import traeger


def perform_operations(scheduler: traeger.Scheduler, mailbox: traeger.Mailbox) -> None:
    for operation, amount in [
        ("deposit", 1000),
        ("deposit", 500),
        ("deposit", 0),
        ("debit", -2000),
        ("debit", 750),
        ("deposit", 250),
        ("debit", 500),
    ]:
        print(f"Performing {operation} {amount}")
        (
            mailbox.send(scheduler, operation, amount)
            .then_result(
                lambda balance_value: print(
                    f"Balance after {operation} is {balance_value}"
                )
            )
            .fail(
                lambda operation_error: print(
                    f"Error performing {operation}: {operation_error}"
                )
            )
        )

        mailbox.send(scheduler, "balance").then_result(
            lambda balance_value: print(f"The Balance is {balance_value}")
        )

File: examples/example-actor-messaging.cpp
// SPDX-License-Identifier: BSL-1.0

#include <iostream>
#include <thread>
#include <chrono>

#include <traeger/actor/Actor.hpp>

void perform_operations(const traeger::Scheduler &scheduler, const traeger::Mailbox &mailbox)
{
    // Use a decorator to enable type deduction in C++
    const auto actor_mailbox = traeger::Actor::Mailbox{mailbox};

    for (auto [operation, amount] : {
             std::make_pair("deposit", 1000),
             std::make_pair("deposit", 500),
             std::make_pair("deposit", 0),
             std::make_pair("debit", -2000),
             std::make_pair("debit", 750),
             std::make_pair("deposit", 250),
             std::make_pair("debit", 500),
         })
    {
        std::cout << "Performing " << operation << " " << amount << std::endl;
        actor_mailbox
            .send(scheduler, operation, amount)
            .then(
                [operation = operation](const traeger::Value &balance_value)
                {
                    std::cout << "Balance after " << operation
                              << " is " << balance_value << std::endl;
                })
            .fail(
                [operation = operation](const traeger::Error &operation_error)
                {
                    std::cout << "Error performing " << operation
                              << ": " << operation_error << std::endl;
                });
    }

    actor_mailbox
        .send(scheduler, "balance")
        .then(
            [](const traeger::Value &balance_value)
            {
                std::cout << "The balance is " << balance_value << std::endl;
            });
}
